### PR TITLE
Improved metadata permissions for apps

### PR DESCRIPTION
**Pull request by korycins**
I want to merge this change because:
- all users with a valid app user token (the UI one, currently everyone with staff status should be able to obtain such a token) can read and write the metadata of the app they are using (including through { app { metadata {...}}} )
- no app user tokens can read or write the private metadata of any apps, including the app they are using
-  every app can read and write its own private metadata (including through { app { privateMetadata {...}}} ) using the app token (the token assigned during app installation)
- Add a new field to the app user token (the UI one) that contains a list of all permissions the user would normally hold (we already list the effective permissions of the current token, we want to tell the app what the user is capable of outside of the app context)

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/app/resolvers.py [hunk 1]
```diff
@@ -44,7 +42,7 @@ def resolve_access_token_for_app_extension(info, root):
     return None
 
 
-@permission_required(AppPermission.MANAGE_APPS)
+# @permission_required(AppPermission.MANAGE_APPS)
```
Comment (L3str4nge) [inline saleor/graphql/app/resolvers.py:45]: Should be removed instead commented I guess

File: saleor/graphql/app/types.py [hunk 2]
```diff
@@ -31,6 +33,20 @@ def has_required_permission(app: models.App, context):
...
+    if auth_token.get("type") == JWT_THIRDPARTY_ACCESS_TYPE:
+        _, app_id = from_global_id_or_error(auth_token["app"], "App")
+    else:
+        app_id = info.context.app.id if info.context.app else None
+    if app_id is not None and int(app_id) == root.id:
+        return True
+    requester = get_user_or_app_from_context(info.context)
+    if not requester.has_perm(AppPermission.MANAGE_APPS):
+        return False
+    return True
```
Comment (L3str4nge) [inline saleor/graphql/app/types.py:47]: `return requester.has_perm(AppPermission.MANAGE_APPS)`

File: saleor/graphql/app/types.py [hunk 3]
```diff
@@ -259,6 +284,33 @@ def __resolve_references(roots: List["App"], info):
...
+            )
+        return ObjectWithMetadata.resolve_metafield(root, info, key=key)
+
+    @staticmethod
+    def resolve_metafields(root: models.App, info, *, keys=None):
+        has_access = has_access_to_app_public_meta(root, info)
+        if not has_access:
+            raise PermissionDenied(
+                permissions=[AppPermission.MANAGE_APPS, AuthorizationFilters.OWNER]
+            )
```
Comment (L3str4nge) [inline saleor/graphql/app/types.py:311]: This block
```
        has_access = has_access_to_app_public_meta(root, info)
        if not has_access:
            raise PermissionDenied(
                permissions=[AppPermission.MANAGE_APPS, AuthorizationFilters.OWNER]
            )
  ```
   Could be moved to separate function.

File: saleor/graphql/schema.graphql [hunk 4]
```diff
@@ -1077,7 +1077,9 @@ type Query {
   ): AppCountableConnection
 
   """
-  Look up an app by ID. If ID is not provided, return the currently authenticated app. Requires one of the following permissions: OWNER, MANAGE_APPS.
+  Look up an app by ID. If ID is not provided, return the currently authenticated app. 
+  
+  Requires one of the following permissions: AuthorizationFilters.AUTHENTICATED_STAFF_USERAuthorizationFilters.AUTHENTICATED_APP. The authenticated app has access to its resources. Fetching different apps requires MANAGE_APPS permission.
```
Comment (fowczarek) [inline saleor/graphql/schema.graphql:1082]: Could we refactor this description? :police_car:

**Review by fowczarek — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

File: saleor/graphql/app/schema.py [hunk 5]
```diff
@@ -63,17 +64,26 @@ class AppQueries(graphene.ObjectType):
...
-    app = graphene.Field(
+    app = PermissionsField(
         App,
         id=graphene.Argument(graphene.ID, description="ID of the app.", required=False),
         description=(
             "Look up an app by ID. If ID is not provided, return the currently "
-            "authenticated app. Requires one of the following permissions: "
-            f"{AuthorizationFilters.OWNER.name}, {AppPermission.MANAGE_APPS.name}."
+            "authenticated app. \n\nRequires one of the following permissions: "
+            f"{AuthorizationFilters.AUTHENTICATED_STAFF_USER.name} "
```
Comment (maarcingebala) [inline saleor/graphql/app/schema.py:77]: Before only staff users with MANAGE_APPS could use this field, now we will allow to any staff that is authenticated, is that correct?

File: saleor/graphql/app/types.py [hunk 6]
```diff
@@ -31,6 +33,26 @@ def has_required_permission(app: models.App, context):
         )
 
 
+def check_permission_for_acess_to_meta(app: models.App, info):
```
Comment (maarcingebala) [inline saleor/graphql/app/types.py:36]: typo: "aces" -> "access"

File: saleor/graphql/app/types.py [hunk 7]
```diff
@@ -183,9 +205,19 @@ class App(ModelObjectType):
     )
     name = graphene.String(description="Name of the app.")
     type = AppTypeEnum(description="Type of the app.")
-    tokens = NonNullList(AppToken, description="Last 4 characters of the tokens.")
+    tokens = NonNullList(
+        AppToken,
+        description=(
+            "Last 4 characters of the tokens. Requires one of the following "
+            f"permissions: {AppPermission.MANAGE_APPS}"
```
Comment (maarcingebala) [inline saleor/graphql/app/types.py:212]: We should also include "OWNER" permission in this description, as owners can also fetch it.

File: saleor/graphql/app/types.py [hunk 8]
```diff
@@ -183,9 +205,19 @@ class App(ModelObjectType):
...
+        ),
+    )
     webhooks = NonNullList(
-        Webhook, description="List of webhooks assigned to this app."
+        Webhook,
+        description=(
+            "List of webhooks assigned to this app. Requires one of the following "
+            f"permissions: {AppPermission.MANAGE_APPS}"
```
Comment (maarcingebala) [inline saleor/graphql/app/types.py:219]: Same here about the owner.

File: saleor/graphql/app/schema.py (hunk 5 above)
Comment (patrys) [reply saleor/graphql/app/schema.py:77]: @maarcingebala Extra space before the `\n` 😄

File: saleor/graphql/app/types.py (hunk 7 above)
Comment (patrys) [reply saleor/graphql/app/types.py:212]: Could we use the same function across the entire core to format the "Requires (...)" message with consistent newlines etc.?

File: saleor/graphql/app/schema.py (hunk 5 above)
Comment (korycins) [reply saleor/graphql/app/schema.py:77]: @maarcingebala Yes, it is expected. We want to allow using the apps by any staff users. The JWT token that is passed to App's iframe will have a separate section with the staff user's permissions, so the app can decide what part of the logic should be available for the user.

File: saleor/graphql/app/types.py [hunk 9]
```diff
@@ -208,15 +212,23 @@ class App(ModelObjectType):
...
+            """
         ),
     )
     webhooks = NonNullList(
         Webhook,
         description=(
-            "List of webhooks assigned to this app. Requires one of the following "
-            f"permissions: {AppPermission.MANAGE_APPS}"
+            f"""List of webhooks assigned to this app.{
+                message_one_of_permissions_required(
```
Comment (patrys) [inline saleor/graphql/app/types.py:227]: Maybe use `+` instead of an `f`-string here?

**Pull request merged**